src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php

```diff
@@ -77,28 +83,29 @@ public function decide(TokenInterface $token, array $attributes, $object = null/
      * If all voters abstained from voting, the decision will be based on the
      * allowIfAllAbstainDecisions property value (defaults to false).
      */
-    private function decideAffirmative(TokenInterface $token, array $attributes, $object = null): bool
+    private function decideAffirmative(TokenInterface $token, array $attributes, $object = null): AccessDecision
     {
+        $votes = [];
         $deny = 0;
         foreach ($this->voters as $voter) {
-            $result = $voter->vote($token, $object, $attributes);
+            $votes[] = $vote = $this->vote($voter, $token, $object, $attributes);
 
-            if (VoterInterface::ACCESS_GRANTED === $result) {
-                return true;
+            if ($vote->isGranted()) {
+                return AccessDecision::createGranted($votes);
             }
 
-            if (VoterInterface::ACCESS_DENIED === $result) {
+            if ($vote->isDenied()) {
                 ++$deny;
-            } elseif (VoterInterface::ACCESS_ABSTAIN !== $result) {
+            } elseif (!$vote->isAbstain()) {
                 trigger_deprecation('symfony/security-core', '5.3', 'Returning "%s" in "%s::vote()" is deprecated, return one of "%s" constants: "ACCESS_GRANTED", "ACCESS_DENIED" or "ACCESS_ABSTAIN".', var_export($result, true), get_debug_type($voter), VoterInterface::class);
```

Comment:
```suggestion
                trigger_deprecation('symfony/security-core', '5.3', 'Returning "%s" in "%s::vote()" is deprecated, return one of "%s" constants: "ACCESS_GRANTED", "ACCESS_DENIED" or "ACCESS_ABSTAIN".', var_export($votes, true), get_debug_type($voter), VoterInterface::class);
```

?

Reply:
this condition needs to be moved down to vote method, still WIP 😄